test/compress/pure_funcs.js

```diff
@@ -772,13 +772,6 @@ inline_pure_call_3: {
         console.log(a);
     }
     expect: {
-        var a = function() {
-            console.log("PASS");
-        }();
-        console.log(a);
+        console.log(void 0);
```

Comment:
The result is incorrect. The pure function return value is used. It should output "PASS" followed by "undefined".

Reply:
There is no return value from the function though, is there?

Reply:
I mean, if they declare this function to be side-effect-free, then why would I care to retain its side effects as long as I retain its behaviour outside of the blackbox? 🤔

Reply:
What do you mean? function f() is returning a value, so the pure call result is used.

Reply:
Let me clarify &minus; the `@__PURE__` function call does not rely on any internal side effects to return a value, so we take that value & discard the rest as it was _declared_ to not have any side effects.

Reply:
Granted, the function in the pure call doesn't explicitly have a return value, but any function without a return implicitly has a `return void 0`. Since each pure call is a black box it has to be treated as if it has a return value to maintain logical consistency.

Reply:
Agreed on the implicit return value. However I'd expect the function call is allowed to be dropped even in the case of say:
```js
var a = /*@__PURE__*/ function() {
    console.log("SKIPPED");
    return 42;
}();
console.log(a);
```
because the user said they don't care about any side effects in there.

Reply:
Think of this call:
`var result = /*@__PURE__*/(function(){ effect(); return 123; })();`
If `result` is used as in `console.log(result)` then we cannot drop `effect()`.
Now just change the `123` in the example above to `void 0`.
`result` is still used, so `effect()` must be retained.

Reply:
It's funny we both used the same example to prove our respective points.

The pure function is a black box. If you use the return value you MUST retain all behaviors WITHIN the black box.

Reply:
You've just stated the opposite of what I said above 😅 

In fact, even without this PR https://github.com/mishoo/UglifyJS/pull/3873#discussion_r422654044 already gets optimised to `console.log(42);`

Reply:
Okay now that we know where we disagree on &minus; care to make an example use case where your preferred behaviour would make this `@__PURE__` feature a showstopper if not followed strictly.

Reply:
You have to trust me on this one. Replace `console.log("SKIPPED")` in your example with a computation that updates a database followed by it returning a value from the database. If that value is used, the database update must take place.

Reply:
In which case we have a bug on our hands, since the assumption is already broken without this PR 😓

Let me try and fix that first before updating this PR then.

Reply:
This is core to how pure calls work.
What assumption is broken without this PR?

Reply:
```js
$ cat test.js
var a = /*@__PURE__*/ function() {
    console.log("PASS");
    return 42;
}();
console.log(a);
```
```sh
$ uglifyjs test.js -bc toplevel
console.log(42);
```

Reply:
It must be a very recent commit. v3.9.2 works as expected.

Reply:
Most likely #3870